src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPolymorphismOptions.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Defines polymorphic configuration for a specified base type.
+    /// </summary>
+    public class JsonPolymorphismOptions
+    {
+        private DerivedTypeList? _derivedTypes;
+        private bool _ignoreUnrecognizedTypeDiscriminators;
+        private JsonUnknownDerivedTypeHandling _unknownDerivedTypeHandling;
+        private string? _typeDiscriminatorPropertyName;
+
+        /// <summary>
+        /// Gets the list of derived types supported in the current polymorphic type configuration.
+        /// </summary>
+        public IList<JsonDerivedType> DerivedTypes => _derivedTypes ??= new(this);
+
+        /// <summary>
+        /// When set to <see langword="true"/>, instructs the serializer to ignore any
+        /// unrecognized type discriminator id's and reverts to the contract of the base type.
+        /// Otherwise, it will fail the deserialization.
+        /// </summary>
+        public bool IgnoreUnrecognizedTypeDiscriminators
+        {
+            get => _ignoreUnrecognizedTypeDiscriminators;
+            set
+            {
+                VerifyMutable();
+                _ignoreUnrecognizedTypeDiscriminators = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the behavior when serializing an undeclared derived runtime type.
+        /// </summary>
+        public JsonUnknownDerivedTypeHandling UnknownDerivedTypeHandling
+        {
+            get => _unknownDerivedTypeHandling;
+            set
+            {
+                VerifyMutable();
+                _unknownDerivedTypeHandling = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a custom type discriminator property name for the polymorhic type.
+        /// Uses the default '$type' property name if left unset.
+        /// </summary>
+        [AllowNull]
+        public string TypeDiscriminatorPropertyName
+        {
+            get => _typeDiscriminatorPropertyName ?? JsonSerializer.TypePropertyName;
+            set
+            {
+                VerifyMutable();
+                _typeDiscriminatorPropertyName = value;
+            }
+        }
+
+        private void VerifyMutable() => DeclaringTypeInfo?.VerifyMutable();
+
+        internal JsonTypeInfo? DeclaringTypeInfo { get; set; }
```

Comment:
consider making this private and only exposing `SetDeclaringTypeInfo` which also handles throwing exception if it's already set

Reply:
I considered initially, but ended up with this approach since it's simpler. Making it another method in this class would also result in a deeper stack trace.